enable single end input

alexg9010 opened the pull request:
This PR is supposed to add a single-end branch of the pipeline, solving #29.

After bootstrapping and configuration, the invocation of ```make integration``` runs the single end test data that is provided with the PR. 

Open Issues:

- [x] try single-end and paired-end in one sample sheet
- [x] get a full working report with lineage plots 
  - may need to test with different SE test data,  since all provided samples are discarded due to low mapping rate(0%)

Commit:
snakefile: remove outdated TODOs

Commit:
ssnakefile: marking rules for which a single end version is needed

Commit:
samplesheet_from_filelist: make script taking input args

Commit:
tests: adding subsampled single-end test files

Commit:
tests: adding se-sample sheet

Commit:
snakefile: WIP version (untested) fastp_se rule

alexg9010 committed:
Makefile.am: add integration tests for single end only

alexg9010 committed:
configure.ac: add viridis R package as dependency

alexg9010 committed:
qc_report_per_sample: fix heading for Mutation locations

alexg9010 committed:
snakefile: define dynamic multiqc input files in function

alexg9010 committed:
snakefile: remove fastqc html files from map_input function

alexg9010 committed:
snakefile: use dynamic input files for multiqc rule

alexg9010 committed:
overviewQC: enable single end processing

alexg9010 committed:
variantreport: rename variants output directory

alexg9010 committed:
variantreport: write sample-specific variants file

alexg9010 committed:
variantreport: force writing of variants csv file

alexg9010 committed:
variantreport: write empty file with header if no signature mutations are found

alexg9010 committed:
snakefile: define variants csv file per sample as output of render_variant_report

alexg9010 committed:
snakefile: add rule to summarize variants csv files

alexg9010 committed:
snakefile: turn summarized variants csv file into required input

fix https://github.com/BIMSBbioinfo/pigx_sarscov2_ww/issues/19

alexg9010 committed:
index.Rmd: test if approved_var_plot contains data before proceeding

alexg9010 committed:
snakefile: keep paired-end fastp trimming rule

alexg9010 committed:
snakefile: provide adapter sequence via settings file

alexg9010 committed:
snakefile: add json file as output to fastp rule

alexg9010 committed:
snakefile: add fastp outputs as required inputs for multiqc

alexg9010 committed:
snakefile: use map_input function to get input for rule bwa_align

alexg9010 committed:
snakefile: add function to determine the fastq file extension

alexg9010 committed:
snakefile: use fastq_ext function to rename fastq.gz suffix to fastqc.html

alexg9010 committed:
snakefile: add single end versions of some fastqc rules

alexg9010 committed:
snakefile: add missing zip file to fastqc rules

alexg9010 committed:
snakefile: add zip files as required multiqc input

alexg9010 committed:
snakefile: let multiqc query files from multiqc_input instead of fastqc
directory

- this includes fastp reports and fastqc reports
- leaves option to include multiqc compatible files in other folders

alexg9010 committed:
overviewQC: formatting

alexg9010 committed:
snakefile: reduce inputs of rule create_overviewQC_table

this can be done since trimming is always performed before alignment, thus coverage summary should be sufficient

alexg9010 committed:
Makefile.am: add tests output directories to clean-local target

alexg9010 committed:
scripts: add new creating_variants_summary_table.R

alexg9010 committed:
test.sh: add new test.sh.in to run the full pipeline

vfschumann commented:
One critical point that needs more polishing ist th[e samtools rule for filter aligned reads](https://github.com/BIMSBbioinfo/pigx_sarscov2_ww/blob/84d6601dbfb028fd1021a82331c616f26c60a0d9/snakefile.py#L297). So far it filters for "proper pairs"(2) which are obviously not present in single-end alignments. 

When se input is given the bash call of this rule has to look like [this](https://github.com/BIMSBbioinfo/pigx_sarscov2_ww/commit/1301698cdd31efba60ca926e054b57891ed8221a). 

I think that could be done with an additional function that works similar to  map_input

alexg9010 committed:
snakefile: update samtools filtering flags

alexg9010 committed:
snakefile: include bam index as ivar input

this was unnotice before, since ivar is automatically indexing bam files

alexg9010 committed:
snakefile: be very explicit to trigger movinf fastqc files

alexg9010 committed:
snakefile: update comment

alexg9010 committed:
overviewQC: move libraries to top

alexg9010 committed:
overviewQC: read sample_sheet without stringsAsFactors

alexg9010 committed:
overviewQC: update read_files() function to add filenames

alexg9010 committed:
overviewQC: update apply_fun_get_read_num to only count when file exists

alexg9010 committed:
overviewQC: rename read_num_raw argument for better understanding

alexg9010 committed:
overviewQC: simplify read counting code

alexg9010 committed:
overviewQC: remove filenames from output file

alexg9010 committed:
tests/samplesheet: add single end files

alexg9010 commented on snakefile.py:
```diff
@@ -294,18 +294,22 @@ def vep_input(args):
 rule samtools_filter_aligned:
     input: os.path.join(MAPPED_READS_DIR, '{sample}_aligned_tmp.sam')
     output: os.path.join(MAPPED_READS_DIR, '{sample}_aligned.bam')
     log: os.path.join(LOG_DIR, 'samtools_filter_aligned_{sample}.log')
```
add mandatory filtering for proper pairs

alexg9010 committed:
snakefile: separate flags for samtools filter

- instead of using the sum of flags

alexg9010 committed:
Makefile.am: add new script creating_variants_summary_table.R

alexg9010 committed:
Revert "test.sh: add new test.sh.in to run the full pipeline"

This reverts commit e91b35f39c7efb94594ddae6c92e94372176a2b6.

alexg9010 committed:
pred_mutation_increase: use str_remove instead of str_split hack

- use of str_split led to unexpected error with empty data

alexg9010 committed:
sample_coverage_score: implement checks and simplify code

- add checks for correct file headers
- simplify aggretion by row using rbindlist()

alexg9010 committed:
snakefile: add conditional 'proper-pair' filter for samtools_filter_aligned

alexg9010 committed:
Revert "Makefile.am: add integration tests for single end only"

This reverts commit 4133f2db464ee4707b76f456d593ddc7000d3c05.

alexg9010 committed:
makefile.am: remove extra single-end output folder from cleaning

alexg9010 committed:
test.sh.in: reword comments and messages

alexg9010 commented:
The pipeline now runs from start till the end independent of whether single or paired-end samples are provided in the sample sheet.  

I am struggling to add the [provided SE sample data](https://github.com/BIMSBbioinfo/pigx_sarscov2_ww/pull/100/commits/41c2e238330c85a30f65378864f1de2197d43dc2) and associated [sample sheet](https://github.com/BIMSBbioinfo/pigx_sarscov2_ww/pull/100/commits/85e1d327bb8380c0788b60e48e502cfaa365094c) to the repo, given that the data was derived from long-reads and the pipeline is optimized for short-read data.

alexg9010 commented:
If you are willing, I would appreciate it if you could do a quick code review @rekado and @vfschumann. 
Thanks

rekado commented on scripts/creating_variants_summary_table.R:
```diff
@@ -0,0 +1,52 @@
+  require(dplyr)
+
+  cat(paste("Summarizing", length(files), "variant files."))
+  # TODO: turn this while loop into a vectorized binary function with Reduce and lapply
```
My thoughts exactly!  The while loop is most unusual.

rekado commented on scripts/creating_variants_summary_table.R:
```diff
@@ -0,0 +1,52 @@
+    if ( file.exists (output_file) ) {
+      df1 <- read.table( output_file,
+                         sep = "\t", header = TRUE, colClasses = "character", check.names = FALSE )
+    } else {
```
This branch is only taken *once*, isn't it?  Would be much clearer to move this outside of the loop.

rekado commented on scripts/creating_variants_summary_table.R:
```diff
@@ -0,0 +1,52 @@
+                                                    colnames(df2)), copy = TRUE)
+    # write to output file
+    write.table(output_df, output_file, sep = "\t",
+          row.names = FALSE, quote = FALSE)
```
This looks like a dupe of the code above.  This could be combined.

rekado commented on scripts/overview_QC_table.R:
```diff
@@ -1,32 +1,30 @@
 library(dplyr)
 library(R.utils)
+require(stringr)
```
require vs library

rekado commented on scripts/overview_QC_table.R:
```diff
@@ -52,22 +52,39 @@ apply_fun_lookup <- function ( sample, sample_sheet.df ){
+     countLines(file) / 4
+  } else {
+    0
+  }
```
I don't like this, but I don't feel strongly enough to suggest an alternative.

rekado commented on scripts/report_scripts/index.Rmd:
```diff
@@ -108,8 +108,12 @@ approved_mut_plot <- df_mut %>% filter( samplename  %in% good_samples.df$samplen
 weights <- weights %>% dplyr::select( c(samplename, total_reads))
-approved_var_plot_location_pooled <- pool_by_weighted_mean(approved_var_plot, weights, "day_location")
 
+if (nrow(approved_var_plot) > 0) { # check if filtered dataframe has actual values
```
I agree with the later comment in the code that says that we should just skip the whole snippet when there are no variants.  Duplicating this condition check everywhere is not great.

rekado commented on scripts/report_scripts/index.Rmd:
```diff
@@ -159,48 +172,52 @@ get_h <- function() {
+  plot_list <- list()
+  for ( i in seq_along(locations)) {
+    plot_name <- locations[i]
+    df <- paste0(locations[i],".df")
```
I think it's bad style to modify df in a loop.  This makes code further down depend on this mutation, which can make it really hard to isolate behavior and test things.  Can we avoid mutating global state like this?

rekado commented on scripts/report_scripts/variantreport_p_sample.Rmd:
````diff
@@ -525,59 +525,68 @@ Called SNVs:
 Variant effect prediction (*Ensemble VEP output*):  
 [Download VEP_out.csv](`r sprintf('data:text/csv;base64,%s', vep)`)
 
-```{r csv_output_variant_plot, include = F, eval = executeDeconvolution}
````
is there any use for executeDeconvolution left?  If not: has it been deleted elsewhere?

alexg9010 replied:
this and the following issues are captured in https://github.com/BIMSBbioinfo/pigx_sars-cov-2/issues/114

alexg9010 replied:
will resolve this for now, since @jonasfreimuth is going to work on this file more extensively

rekado commented on scripts/report_scripts/variantreport_p_sample.Rmd:
```diff
@@ -525,59 +525,68 @@ Called SNVs:
-        output_variant_plot[sample_row,][i] <- df$value[df$name == i]
-        output_variant_plot <- output_variant_plot %>% mutate(others = 1-rowSums(across(all_of(all_variants)), na.rm = TRUE))
-}
+if (!executeDeconvolution) {
```
oh, answered my question from above.

Please invert the if condition.  When using if / else always aim to use a positive condition.

rekado commented on scripts/report_scripts/variantreport_p_sample.Rmd:
```diff
@@ -525,59 +525,68 @@ Called SNVs:
-        output_variant_plot <- output_variant_plot %>% mutate(others = 1-rowSums(across(all_of(all_variants)), na.rm = TRUE))
-}
+if (!executeDeconvolution) {
+  # if no signatur mutation found write empty output file
```
typo: "signatur"

rekado commented on scripts/report_scripts/variantreport_p_sample.Rmd:
```diff
@@ -525,59 +525,68 @@ Called SNVs:
+          output_variant_plot[sample_row,][i] <- df$value[df$name == i]
+          output_variant_plot <- output_variant_plot %>% mutate(others = 1-rowSums(across(all_of(all_variants)), na.rm = TRUE))
+  }
+    }
```
weird indentation.

rekado commented on scripts/report_scripts/variantreport_p_sample.Rmd:
```diff
@@ -525,59 +525,68 @@ Called SNVs:
+
+  # write mutation frequency values to df
+  for (i in all_variants){
+    if (i %in% df$name){
```
shouldn't we filter all_variants instead of having a one-branch if here?

rekado commented on scripts/report_scripts/variantreport_p_sample.Rmd:
```diff
@@ -525,59 +525,68 @@ Called SNVs:
+  for (i in all_variants){
+    if (i %in% df$name){
+      # check if variant already has a column
+      if (i %in% colnames(output_variant_plot)) {
```
shouldn't this condition be merged with the above?  This would reduce unnecessary nesting.

rekado commented on scripts/report_scripts/variantreport_p_sample.Rmd:
```diff
@@ -525,59 +525,68 @@ Called SNVs:
-  output_variant_plot <- as.data.frame(lapply(output_variant_plot, as.character), check.names = FALSE)
-  # merge with adding cols and rows
-  output_variant_plot <- full_join(previous_df, output_variant_plot, by = colnames(previous_df), copy = TRUE)
+  ## # TODO: This chunk hast to go into a seperate rule
```
Let's delete this instead of commenting it.  It's either needed or not.  Let's not clutter this already cluttered file :)

rekado commented on snakefile.py:
```diff
@@ -170,18 +170,46 @@ def trim_reads_input(args):
+            expand(os.path.join(FASTQC_DIR, '{sample}', '{sample}_trimmed_fastqc.zip'), sample=sample),
+            expand(os.path.join(FASTQC_DIR, '{sample}', '{sample}_aligned_sorted_primer-trimmed_sorted_fastqc.html'), sample = sample),
+            expand(os.path.join(FASTQC_DIR, '{sample}', '{sample}_aligned_sorted_primer-trimmed_sorted_fastqc.zip'), sample = sample)
+        ]
```
I'm a bit sad to see this much duplication.  I see why you do it, but there's got to be a better way, especially since read_num is predictably the same.  I'm sure we can avoid this and generate the `files` array by mapping over reads_files instead of having this ugly `if` condition.

rekado commented on snakefile.py:
```diff
@@ -217,21 +242,32 @@ def vep_input(args):
+        {FASTP_EXEC} -i {input[0]} -o {output.r} --html {output.html} --json {output.json} >> {log}t 2>&1
     """
-    
+# !!! 03/01 Not tested if the report output works like that
```
oh?

rekado commented on snakefile.py:
```diff
@@ -321,24 +395,34 @@ def vep_input(args):
                     mv {params.output_dir}/{tmp_R2_zip} {output.r2_zip}
                 fi """)
  
 # TODO: can probably be done by using map_input, no seperate functions neccessary?        
```
what does this comment refer to?  Is it still valid?

alexg9010 replied:
I think this comment is still valid, but I won't resolve it now.

rekado commented on snakefile.py:
```diff
@@ -445,7 +527,7 @@ def no_variant_vep(sample, lofreq_output):
     shell: "{IMPORT_TAXONOMY_EXEC} -m 3 -t 5 {input.kraken_output} -tax {input.database} -o {output} >> {log} 2>&1"
 
-
+# TODO: change amplicon naming to mutaiton site bc it's missleading
```
typos.  "mutation" and "misleading".  And "because" ;)

rekado commented on snakefile.py:
```diff
@@ -464,7 +546,7 @@ def no_variant_vep(sample, lofreq_output):
     shell: "{SAMTOOLS_EXEC} coverage {input.aligned_bam} > {output} 2>> {log} 3>&2"
 
-
+# TODO: change amplicon naming to mutation site bc it's missleading
```
fewer typos, but still some :)

rekado reviewed (changes requested):
Thank you for this huge patch set!  I've got a couple of nitpicks.  Maybe we can address some of them before merging.

alexg9010 committed:
creating_variants_summary_table.R: rewrite while loop

into a vectorized binary function with Reduce and lapply

alexg9010 committed:
creating_variants_summary_table.R: extract writing of output

from create_summary

alexg9010 committed:
creating_mutation_summary_table.R: rewrite while loop

into vectorized binary function

alexg9010 committed:
creating_mutation_summary_table.R: extract output-writing

from  create_summary function

alexg9010 committed:
overiew_QC_table.R: put library instead of require

alexg9010 committed:
.gitignore: ignore .Rproj.user file

alexg9010 committed:
snakefile: remove duplication in multiqc_input

by extracting templates and fillling dependent wildcards

alexg9010 committed:
snakefile: remove comment since pipeline was tested

alexg9010 committed:
snakefile: remove comment since parameters can be passed via settings file

alexg9010 committed:
snakefile: delete trailing whitespaces

alexg9010 committed:
snakefile: correct comments

alexg9010 committed:
overviewQC: rewrite read_files() and apply_fun_lookup() as parse_sample_sheet()

join both functions, since splitting and joining the sample sheet by rows is not
necessary for simple column-wise manipulations

alexg9010 committed:
overviewQC: be more explicit about required folder paths

alexg9010 committed:
overviewQC: extend function for parsing of sample sheet

- define and save file paths
- works for both paired and single-end

alexg9010 committed:
overviewQC: rewrite read_num_raw as vectorized  function

- remove function apply_fun_get_read_num

alexg9010 committed:
overviewQC:  rewrite parse_amplicons as parse_amplicon_coverage

- merge with apply_fun_parse_coverage_file
- rewrite as vectorized function

alexg9010 committed:
overviewQC: simplify and format code

alexg9010 committed:
overviewQC: cosmetic fixup

alexg9010 committed:
overViewQC:  fix wrong argument number

alexg9010 committed:
snakefile: fix ordering of folder paths given to overviewQC script

alexg9010 commented:
@rekado and @vfschumann 
You may review the last time changes I made to the overViewQC script. 
This PR has passed `make distcheck` and can be safely merged.

alexg9010 committed:
pooling.R: replace multiple require calls with initial library call

alexg9010 committed:
pred_mutation_increase.R: replace require calls with library calls and replace
tidyverse  with dplyr

alexg9010 committed:
table_extraction: summarize require calls in single library call

alexg9010 committed:
index.Rmd: set default chunk options

alexg9010 committed:
index.Rmd: introduce VARIANTS_FOUND to skip snippets

- conditionally skip code chunks and text blocks

alexg9010 committed:
index.Rmd: split chunk 'libraries_and_input'

- create seperate chunks 'libraries', 'input', 'functions'

alexg9010 committed:
index.Rmd: write function to create html formatted notes

alexg9010 committed:
index.Rmd: format error, warning and messages with html styles

alexg9010 committed:
index.Rmd:  conditional printing of Lineage-abundance geo-localized section

- replace html notice container with create_html_note call
- only print slider text if more than one date is shown

alexg9010 committed:
index.Rmd: print description of mutation notation as html note

alexg9010 committed:
index.Rmd: remove unused color palette

- was probably replaced by viridis color palette back then

alexg9010 committed:
index.Rmd: add information how to use notifications

alexg9010 committed:
index.Rmd: conditionally show mutation-related chunks

alexg9010 committed:
index.Rmd: introduce global variables RUN_MUTATION_REGRESSION and
APPROVED_VAR_PLOT

- used to conditionally evaluate chunks related to regression analysis

alexg9010 committed:
index.Rmd: add function to dump data in csv format

alexg9010 committed:
index.Rmd: add function to create link with named output file

alexg9010 committed:
index.Rmd: simplify chunk headers

alexg9010 committed:
index.Rmd: update file download links

- use function create_html_download_link instead of markdown links

alexg9010 committed:
index.Rmd: simplify with dplyr

alexg9010 committed:
index.Rmd: turn cat into warning

alexg9010 committed:
index.Rmd: fix formatting issues

alexg9010 committed:
index.Rmd: update xaxis title in p_variant_pooled

alexg9010 committed:
index.Rmd: update slider title in maps_for_variants when applicable

alexg9010 committed:
index.Rmd: conditionally evaluate linear regression chunks

- only eval if RUN_MUTATION_REGRESSION is TRUE
- update APPROVED_MUTATIONS_FOUND if significant mutations are found
- only show regression results (text and table) if something was found otherwise print warning
- update download links

alexg9010 committed:
index.Rmd: conditionally mutation trend chunks

alexg9010 committed:
index.Rmd: fix typo

alexg9010 committed:
index.Rmd: turn text for trending mutations geo-localized into code chunks

alexg9010 committed:
index.Rmd: Conditionally run mutations_for_locations_and_date chunk

alexg9010 committed:
index.Rmd: insert empty line before headers

alexg9010 committed:
index.Rmd: update location_plot function to include melted as argument

alexg9010 committed:
index.Rmd: rewrite plot_list for loops using lapply

alexg9010 committed:
set mutation coverage threshold via settings file
